crates/compiler/src/obj_gen/mod.rs

```diff
@@ -0,0 +1,1132 @@
+
+        // This instruction takes two forms. The first takes one operand, which is the
+        // value to return, while the other takes no operands, and just causes control
+        // flow.
+        let operands = instruction.get_operands().collect_vec();
+        let return_ids = match operands.len() {
+            0 => {
+                // When there are no operands, this is pure control flow so we return no values.
+                Vec::new()
+            }
+            1 => {
+                // When there is one operand, we are returning a specific value back.
+                let value = extract_value_operand(operands[0], InstructionOpcode::Return)?;
+                let value_name = value.get_name().to_str()?;
+                let value_type = LLVMType::try_from(value.get_type())?;
+                let expected_return_type = &func_ctx.typ().return_type;
+                assert_eq!(expected_return_type.as_ref(), &value_type);
+
+                let return_value_id = func_ctx.lookup_variable(value_name).unwrap_or_else(|| {
+                    panic!("Returned variable {value_name} does not exist in function context")
+                });
+
+                vec![return_value_id]
+            }
+            _ => Err(Error::malformed_llvm(&format!(
+                "ret instruction encountered with {len} operands where only 0 or 1 are supported",
+                len = operands.len()
+            )))?,
+        };
+
+        bb.end_with_return(return_ids);
+
+        Ok(())
+    }
+
+    /// Generates the FLO code that corresponds to an LLVM
+    /// [br](https://llvm.org/docs/LangRef.html#br-instruction) instruction.
+    ///
+    /// # Errors
+    ///
+    /// - [`Error`] if the correct FLO representation of the instruction cannot
+    ///   be generated.
+    ///
+    /// # Panics
+    ///
+    /// - If the provided instruction is _not_ a `br`.
+    /// - If the instruction has the incorrect number of operands.
+    /// - If any target block is unknown.
+    /// - If the condition to the conditional branch is not a boolean.
+    pub fn generate_br(
+        &self,
+        instruction: InstructionValue,
+        bb: &mut BlockBuilder,
+        func_ctx: &mut FunctionContext,
+    ) -> Result<()> {
+        assert!(
+            matches!(instruction.get_opcode(), InstructionOpcode::Br),
+            "The argument to generate_br was not a br instruction"
+        );
+
+        // First we have to grab the operands to the instruction. There are two forms of
+        // the branch instruction which can be distinguished by the type of the first
+        // operand. If the first operand is a basic block, this is an unconditional
+        // branch, and if it is a value (an i1 used for the condition), it is a
+        // conditional branch.
+        let operands = instruction.get_operands().collect_vec();
+        match operands.len() {
+            1 => {
+                // In this case, we have to generate the UNCONDITIONAL branch.
```

Comment:
would love to see these inner blocks broken out into functions (`generate_unconditional_br`, `generate_conditional_br`)